Mathlib/Order/WellQuasiOrder.lean

```diff
@@ -0,0 +1,151 @@
+/-
+Copyright (c) 2025 Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Violeta Hernández Palacios, Aaron Anderson
+-/
+import Mathlib.Data.Fintype.Card
+import Mathlib.Data.Set.Finite.Basic
+import Mathlib.Order.Antichain
+import Mathlib.Order.OrderIsoNat
+
+/-!
+# Well quasi-orders
+
+A well quasi-order (WQO) is a relation such that any infinite sequence contains an infinite
+subsequence of related elements. For a preorder, this is equivalent to having a well-founded order
+with no infinite antichains.
+
+## Main definitions
+
+* `WellQuasiOrdered`: a predicate for WQO unbundled relations
+* `WellQuasiOrderedLE`: a typeclass for a bundled WQO `≤` relation
+
+## TODO
+
+* Define `Set.PartiallyWellOrderedOn` and `Set.IsPWO` in terms of these predicates, and rename them
+  to match.
```

Comment:
This definition has many names in the literature, so we should add tags here to help searchability, here's a quick list off the top of my head, feel free to add more.

wqo, welll quasi-order, partial well order, dickson order